Moonshot/procedural_map_generation/Room.gd

```diff
@@ -93,5 +104,102 @@ func is_instanced():
+	if setup_room_bool:
+		get_spawn_nodes()
+		spawn_enemies()
+		spawn_items()
 	room_instanced = true
 	
+# Function to select a random level setup and remove the others from the room
+func select_level_setup():
+	# Retrieve the different level setups
+	var level_setups_node = room_instance.get_node('level_setups')
+	if level_setups_node == null:
+		return false
+	var level_setups = level_setups_node.get_children()
+	selected_level_setup = level_setups[randi()%level_setups.size()]
+	# Delete all the level setups not nee
+	for level_setup in level_setups:
+		if not level_setup.name == selected_level_setup.name:
+			level_setup.queue_free()
+	return true
+
+# Fetches all the spawn nodes and stores them in the class for easier access
+func get_spawn_nodes():
+	var spawn_nodes = selected_level_setup.get_children()
+	for node in spawn_nodes:
+		if node.name.begins_with('baddie_spawn_zone'):
+			baddie_spawn_areas.append(node)
+		elif node.name.begins_with('item_spawn_point'):
+			item_spawn_nodes.append(node)
+		elif node.name.begins_with('loot_destroyable'):
+			loot_destroyable_objects.append(node)
+			
+			
+# Spawns a number of enemies based on the level and size of the room
+func spawn_enemies():
+	room_size_x = (room_instance.get_node('camera_limit_SE').position.x - room_instance.get_node('camera_limit_NW').position.x)
+	room_size_y = (room_instance.get_node('camera_limit_SE').position.y - room_instance.get_node('camera_limit_NW').position.y)
+	var difficulty_rating = room_size_x*room_size_y*(pow(level_no, scaling_factor))
+	var no_enemies = rand_range(0.3, 1) * difficulty_rating / 60000
+	# Retrieve list of enemies and attributes of enemies (spawn on ground/air/wall + size + difficulty)
+	var baddie_list = get_baddie_list()
+	
+	# Select one type of enemie to include in the map
+	var selected_baddie = baddie_list[randi()%baddie_list.size()]
+	var baddie_scene = load("res://baddies/" + selected_baddie)
+	
+	# Spawn the enemies, initally randomly and then later on walls/floor etc.
+	var level_setup_nodes = selected_level_setup.get_children()
+	
+	# Split the enemies across all the spawn areas
+	for i in range(no_enemies):
+		# Select a spawn area at random
+		var spawn_area = baddie_spawn_areas[randi()%baddie_spawn_areas.size()].polygon
+		var max_x = -100000
+		var min_x = 100000
+		var max_y = -100000
+		var min_y = 100000
+		# Get the limits of the spawn area (max & min x, y across all points)
+		for point in spawn_area:
+			if point.x > max_x:
+				max_x = point.x
+			elif point.x < min_x:
+				min_x = point.x
+			if point.y > max_y:
+				max_y = point.y
+			elif point.y < min_y:
+				min_y = point.y
+				
+		# Select a random position within the spawn area 
+		var rand_x = rand_range(min_x,max_x)
+		var rand_y = rand_range(min_y, max_y)
+		
+		# Spawn baddie
+		var new_baddie = baddie_scene.instance()
+		new_baddie.position.x = rand_x
+		new_baddie.position.y = rand_y
+		room_instance.add_child(new_baddie)
+		
```

Comment:
I'm really wary of this. I don't like the idea of enemy spawning being handled by the room gen in the room class. 

Additionally random enemy generation makes room design tricky.

Reply:
So currently you have to option to either include enemies directly in a level setup or to use the spawn areas. The third option is to have multiple spawn points, of which only some are used. This would mean less code, but it would also mean that theres less control over the difficulty of the room based on the level without creating additional level setups.

What is the issue of spawing enemies into the room from the Room class? As it holds the active instance for that room and does the instancing. Is there another place we could put it?

Reply:
I don't believe difficulty can be boiled down to a number that we can control. Had chosen enemies combined with bespoke room design to enhance each other is where difficulty and challenge can arrise. So from that I think enemy spawning should be controlled from within the room scene, either on ready or through player interaction.

 The waryness of the room class for me is how tied up it is to the room gen. If we want to build a room and test it, wouldn't it currently be impossible to test without running the full game and finding the room? As simply running the room would mean no "room class" is generated for that room and in turn enemy spawning would not function maybe even mean the scene would crash?